services/shhext/publisher/service.go

```diff
@@ -0,0 +1,446 @@
+	"os"
+	"path/filepath"
+	"time"
+)
+
+const (
+	tickerInterval   = 120
+	maxInstallations = 3
+)
+
+var (
+	errProtocolNotInitialized = errors.New("procotol is not initialized")
+	// ErrPFSNotEnabled is returned when an endpoint PFS only is called but
+	// PFS is disabled
+	ErrPFSNotEnabled = errors.New("pfs not enabled")
+)
+
+//type Persistence interface {
+//}
+
+type Service struct {
+	whisper    *whisper.Whisper
+	whisperAPI *whisper.PublicWhisperAPI
+	protocol   *chat.ProtocolService
+	//	persistence Persistence
+	log    log.Logger
+	filter *filter.Service
+	config *Config
+	quit   chan struct{}
+	ticker *time.Ticker
+}
+
+type Config struct {
+	PfsEnabled     bool
+	DataDir        string
+	InstallationID string
+}
+
+func New(config *Config, w *whisper.Whisper) *Service {
+	return &Service{
+		config:     config,
+		whisper:    w,
+		whisperAPI: whisper.NewPublicWhisperAPI(w),
+		log:        log.New("package", "status-go/services/publisher.Service"),
+	}
+}
+
+// InitProtocolWithPassword creates an instance of ProtocolService given an address and password used to generate an encryption key.
+func (s *Service) InitProtocolWithPassword(address string, password string) error {
+	digest := sha3.Sum256([]byte(password))
+	encKey := fmt.Sprintf("%x", digest)
+	return s.initProtocol(address, encKey, password)
+}
+
+// InitProtocolWithEncyptionKey creates an instance of ProtocolService given an address and encryption key.
+func (s *Service) InitProtocolWithEncyptionKey(address string, encKey string) error {
+	return s.initProtocol(address, encKey, "")
+}
+
+func (s *Service) initProtocol(address, encKey, password string) error {
+	if !s.config.PfsEnabled {
+		return nil
+	}
+
+	if err := os.MkdirAll(filepath.Clean(s.config.DataDir), os.ModePerm); err != nil {
+		return err
+	}
+	v0Path := filepath.Join(s.config.DataDir, fmt.Sprintf("%x.db", address))
+	v1Path := filepath.Join(s.config.DataDir, fmt.Sprintf("%s.db", s.config.InstallationID))
+	v2Path := filepath.Join(s.config.DataDir, fmt.Sprintf("%s.v2.db", s.config.InstallationID))
+	v3Path := filepath.Join(s.config.DataDir, fmt.Sprintf("%s.v3.db", s.config.InstallationID))
+	v4Path := filepath.Join(s.config.DataDir, fmt.Sprintf("%s.v4.db", s.config.InstallationID))
+
+	if password != "" {
+		if err := appDB.MigrateDBFile(v0Path, v1Path, "ON", password); err != nil {
+			return err
+		}
+
+		if err := appDB.MigrateDBFile(v1Path, v2Path, password, encKey); err != nil {
+			// Remove db file as created with a blank password and never used,
+			// and there's no need to rekey in this case
+			os.Remove(v1Path)
+			os.Remove(v2Path)
+		}
+	}
+
+	if err := appDB.MigrateDBKeyKdfIterations(v2Path, v3Path, encKey); err != nil {
+		os.Remove(v2Path)
+		os.Remove(v3Path)
+	}
+
+	// Fix IOS not encrypting database
+	if err := appDB.EncryptDatabase(v3Path, v4Path, encKey); err != nil {
+		os.Remove(v3Path)
+		os.Remove(v4Path)
+	}
+
+	// Desktop was passing a network dependent directory, which meant that
```

Comment:
shouldn't that be fixed in Desktop instead? looks off to me

Reply:
cc @vkjr

Reply:
The code has just been moved, it was there before this PR, keeping in desktop would have been sensible, but decided otherwise in order to have a single place to handle migrations, eventually we can get rid of this (0.9.0 had this defect)